### PR TITLE
fix deadlock in job-exporter by disabling GCCollector

### LINKED ISSUE #2104: job-exporter deadlock problem
We found job-exporter has deadlock problem, this will make job-exporter can not export metrics to outside world, the stack trace of main thread is:
```
#0  0x00007f60418a9536 in futex_abstimed_wait_cancelable (private=0,
    abstime=0x0, expected=0, futex_word=0x556a3ca26480)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  do_futex_wait (sem=sem@entry=0x556a3ca26480, abstime=0x0)
    at sem_waitcommon.c:111
#2  0x00007f60418a95e4 in __new_sem_wait_slow (sem=0x556a3ca26480, abstime=0x0)
    at sem_waitcommon.c:181
#3  0x00007f60418a9679 in __new_sem_wait (sem=sem@entry=0x556a3ca26480)
    at sem_wait.c:29
#4  0x00007f6041c7b828 in PyThread_acquire_lock_timed (
    lock=lock@entry=0x556a3ca26480, microseconds=microseconds@entry=-1000000,
    intr_flag=intr_flag@entry=1) at Python/thread_pthread.h:340
#5  0x00007f6041cc8d5a in acquire_timed (timeout=-1000000000,
    lock=0x556a3ca26480) at ./Modules/_threadmodule.c:61
#6  lock_PyThread_acquire_lock (self=0x7f603d66bc38, args=<optimized out>,
    kwds=<optimized out>) at ./Modules/_threadmodule.c:144
#7  0x00007f6041b50006 in _PyMethodDef_RawFastCallDict (
    method=0x7f6041fe1f40 <lock_methods+192>, self=<optimized out>, args=0x0,
    nargs=0, kwargs=<optimized out>) at Objects/call.c:515
#8  0x00007f6041b50285 in _PyCFunction_FastCallDict (func=0x7f603cce4dc8,
    args=<optimized out>, nargs=<optimized out>, kwargs=<optimized out>)
    at Objects/call.c:582
#9  0x00007f6041b248a0 in _PyEval_EvalFrameDefault (f=<optimized out>,
    throwflag=<optimized out>) at Python/ceval.c:2899
#10 0x00007f6041c3bb33 in _PyEval_EvalCodeWithName (_co=0x7f603d6e80c0,
    globals=<optimized out>, locals=locals@entry=0x0, args=<optimized out>,
    argcount=2, kwnames=0x0, kwargs=0x556a3cb5c160, kwcount=0, kwstep=1,
    defs=0x0, defcount=0, kwdefs=0x0, closure=0x0, name=0x7f6040c6b228,
    qualname=0x7f603d62ee40) at Python/ceval.c:3923
#11 0x00007f6041b4f5e6 in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:433
#12 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#13 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#14 0x00007f6041c3bb33 in _PyEval_EvalCodeWithName (
    _co=_co@entry=0x7f603d646420, globals=globals@entry=0x7f603d638ca8,
    locals=locals@entry=0x0, args=args@entry=0x7ffd6d01fee0,
    argcount=argcount@entry=2, kwnames=kwnames@entry=0x0, kwargs=0x0,
    kwcount=0, kwstep=2, defs=0x0, defcount=0, kwdefs=0x0,
    closure=0x7f603d6a7410, name=0x7f60409b3570, qualname=0x7f603d691be0)
    at Python/ceval.c:3923
#15 0x00007f6041b4f337 in _PyFunction_FastCallDict (func=0x7f603cfd0510,
    args=0x7ffd6d01fee0, nargs=2, kwargs=0x0) at Objects/call.c:376
#16 0x00007f6041b50676 in _PyObject_CallFunctionVa (
    callable=callable@entry=0x7f603cfd0510,
    format=format@entry=0x7f6041d222e4 "sO", va=va@entry=0x7ffd6d01ff58,
    is_size_t=is_size_t@entry=0) at Objects/call.c:957
#17 0x00007f6041b50eab in _PyObject_CallFunctionVa (is_size_t=0,
    va=0x7ffd6d01ff58, format=0x7f6041d222e4 "sO", callable=0x7f603cfd0510)
    at Objects/call.c:930
#18 PyObject_CallFunction (callable=callable@entry=0x7f603cfd0510,
    format=format@entry=0x7f6041d222e4 "sO") at Objects/call.c:977
#19 0x00007f6041c910b1 in invoke_gc_callback (
    phase=phase@entry=0x7f6041d18c5f "stop", generation=generation@entry=0,
    collected=<optimized out>, uncollectable=<optimized out>)
    at Modules/gcmodule.c:1008
#20 0x00007f6041c91870 in collect_with_callback (generation=0)
    at Modules/gcmodule.c:1029
#21 collect_generations () at Modules/gcmodule.c:1051
#22 0x00007f6041c91f11 in _PyObject_GC_Alloc (basicsize=<optimized out>,
    use_calloc=0) at Modules/gcmodule.c:1708
#23 _PyObject_GC_Malloc (basicsize=<optimized out>) at Modules/gcmodule.c:1718
#24 0x00007f6041c91fed in _PyObject_GC_New (
    tp=tp@entry=0x7f6041f9e760 <PyDict_Type>) at Modules/gcmodule.c:1730
#25 0x00007f6041b948d3 in new_dict (values=0x0, keys=0x7f603c629768)
    at Objects/dictobject.c:582
#26 clone_combined_dict (orig=0x7f603cfacb40) at Objects/dictobject.c:647
#27 PyDict_Copy (o=0x7f603cfacb40) at Objects/dictobject.c:2581
#28 0x00007f6041b4fad0 in _PyMethodDef_RawFastCallKeywords (
    method=0x7f6041f9dbe0 <mapp_methods+416>, self=self@entry=0x7f603cfacb40,
    args=args@entry=0x556a3cb6d1c8, nargs=nargs@entry=0,
    kwnames=kwnames@entry=0x0) at Objects/call.c:629
#29 0x00007f6041b581c9 in _PyMethodDescr_FastCallKeywords (
    descrobj=0x7f60420500d8, args=args@entry=0x556a3cb6d1c0,
    nargs=nargs@entry=1, kwnames=kwnames@entry=0x0)
    at Objects/descrobject.c:288
#30 0x00007f6041b28eba in call_function (kwnames=0x0, oparg=1,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4563
#31 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#32 0x00007f6041b61d83 in gen_send_ex (closing=0, exc=0, arg=0x0,
    gen=0x7f603ccccde0) at Objects/genobject.c:221
#33 gen_iternext (gen=0x7f603ccccde0) at Objects/genobject.c:542
#34 0x00007f6041b21681 in _PyEval_EvalFrameDefault (f=<optimized out>,
    throwflag=<optimized out>) at Python/ceval.c:2802
#35 0x00007f6041b1f5b0 in function_code_fastcall (co=<optimized out>,
    args=<optimized out>, nargs=1, globals=<optimized out>)
    at Objects/call.c:283
#36 0x00007f6041b4f6bf in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:415
#37 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#38 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#39 0x00007f6041b61d83 in gen_send_ex (closing=0, exc=0, arg=0x0,
    gen=0x7f603d673f48) at Objects/genobject.c:221
#40 gen_iternext (gen=0x7f603d673f48) at Objects/genobject.c:542
#41 0x00007f6041b21681 in _PyEval_EvalFrameDefault (f=<optimized out>,
    throwflag=<optimized out>) at Python/ceval.c:2802
#42 0x00007f6041c3bb33 in _PyEval_EvalCodeWithName (_co=0x7f603d68b4b0,
    globals=<optimized out>, locals=locals@entry=0x0, args=<optimized out>,
    argcount=1, kwnames=0x0, kwargs=0x7f603c620e10, kwcount=0, kwstep=1,
    defs=0x7f603d6a1a38, defcount=1, kwdefs=0x0, closure=0x0,
    name=0x7f603d6de430, qualname=0x7f603d6de430) at Python/ceval.c:3923
#43 0x00007f6041b4f5e6 in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:433
#44 0x00007f6041b280e1 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#45 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3117
#46 0x00007f6041c3bb33 in _PyEval_EvalCodeWithName (_co=0x7f603d62dc90,
    globals=<optimized out>, locals=locals@entry=0x0, args=<optimized out>,
    argcount=2, kwnames=0x0, kwargs=0x7f60409473b8, kwcount=0, kwstep=1,
    defs=0x0, defcount=0, kwdefs=0x0, closure=0x7f603cf6d748,
    name=0x7f603d68afb0, qualname=0x7f603d6911e8) at Python/ceval.c:3923
#47 0x00007f6041b4f5e6 in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:433
#48 0x00007f6041b280e1 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#49 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3117
#50 0x00007f6041b1f5b0 in function_code_fastcall (co=<optimized out>,
    args=<optimized out>, nargs=2, globals=<optimized out>)
    at Objects/call.c:283
#51 0x00007f6041b4f6bf in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:415
#52 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#53 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#54 0x00007f6041b1f5b0 in function_code_fastcall (co=<optimized out>,
    args=<optimized out>, nargs=1, globals=<optimized out>)
    at Objects/call.c:283
#55 0x00007f6041b4f6bf in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:415
#56 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#57 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#58 0x00007f6041b1f5b0 in function_code_fastcall (co=co@entry=0x7f603e368780,
    args=<optimized out>, args@entry=0x7ffd6d021080, nargs=nargs@entry=4,
    globals=globals@entry=0x7f603e33d3a8) at Objects/call.c:283
#59 0x00007f6041b4f526 in _PyFunction_FastCallDict (func=0x7f603e36de18,
    args=0x7ffd6d021080, nargs=4, kwargs=0x0) at Objects/call.c:322
#60 0x00007f6041b505fd in _PyObject_Call_Prepend (
    callable=callable@entry=0x7f603e36de18, obj=obj@entry=0x7f603c7ef2b0,
    args=args@entry=0x7f603cce41b0, kwargs=kwargs@entry=0x0)
    at Objects/call.c:904
#61 0x00007f6041bbc889 in slot_tp_init (self=0x7f603c7ef2b0,
    args=0x7f603cce41b0, kwds=0x0) at Objects/typeobject.c:6597
#62 0x00007f6041bb73b2 in type_call (type=<optimized out>,
    args=0x7f603cce41b0, kwds=0x0) at Objects/typeobject.c:949
#63 0x00007f6041b4fc84 in _PyObject_FastCallKeywords (callable=0x556a3cb09e58,
    stack=<optimized out>, nargs=<optimized out>, kwnames=0x0)
    at Objects/call.c:199
#64 0x00007f6041b2698a in call_function (kwnames=0x0, oparg=3,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4589
#65 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3086
#66 0x00007f6041b1f5b0 in function_code_fastcall (co=<optimized out>,
    args=<optimized out>, nargs=3, globals=<optimized out>)
    at Objects/call.c:283
#67 0x00007f6041b4f6bf in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:415
#68 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#69 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#70 0x00007f6041b1f5b0 in function_code_fastcall (co=<optimized out>,
    args=<optimized out>, nargs=3, globals=<optimized out>)
    at Objects/call.c:283
#71 0x00007f6041b4f6bf in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:415
#72 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#73 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#74 0x00007f6041b1f5b0 in function_code_fastcall (co=<optimized out>,
    args=<optimized out>, nargs=1, globals=<optimized out>)
    at Objects/call.c:283
#75 0x00007f6041b4f6bf in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:415
#76 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#77 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#78 0x00007f6041c3bb33 in _PyEval_EvalCodeWithName (_co=0x7f60405e85d0,
    globals=<optimized out>, locals=locals@entry=0x0, args=<optimized out>,
    argcount=1, kwnames=0x0, kwargs=0x556a3ca68ea0, kwcount=0, kwstep=1,
    defs=0x7f603e35c300, defcount=1, kwdefs=0x0, closure=0x0,
    name=0x7f6040c537b0, qualname=0x7f603e360620) at Python/ceval.c:3923
#79 0x00007f6041b4f5e6 in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:433
#80 0x00007f6041b29b17 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#81 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3103
#82 0x00007f6041b1f5b0 in function_code_fastcall (co=<optimized out>,
    args=<optimized out>, nargs=1, globals=<optimized out>)
    at Objects/call.c:283
#83 0x00007f6041b4f6bf in _PyFunction_FastCallKeywords (func=<optimized out>,
    stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at Objects/call.c:415
#84 0x00007f6041b280e1 in call_function (kwnames=0x0, oparg=<optimized out>,
    pp_stack=<synthetic pointer>) at Python/ceval.c:4586
#85 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3117
#86 0x00007f6041c3bb33 in _PyEval_EvalCodeWithName (
    _co=_co@entry=0x7f6040d0e5d0, globals=globals@entry=0x7f6040d311f8,
    locals=locals@entry=0x7f6040d311f8, args=args@entry=0x0,
    argcount=argcount@entry=0, kwnames=kwnames@entry=0x0, kwargs=0x0,
    kwcount=0, kwstep=2, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0,
    name=0x0, qualname=0x0) at Python/ceval.c:3923
#87 0x00007f6041c3bc8e in PyEval_EvalCodeEx (_co=_co@entry=0x7f6040d0e5d0,
    globals=globals@entry=0x7f6040d311f8, locals=locals@entry=0x7f6040d311f8,
    args=args@entry=0x0, argcount=argcount@entry=0, kws=kws@entry=0x0,
    kwcount=0, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0)
    at Python/ceval.c:3952
#88 0x00007f6041c3bcbb in PyEval_EvalCode (co=co@entry=0x7f6040d0e5d0,
    globals=globals@entry=0x7f6040d311f8, locals=locals@entry=0x7f6040d311f8)
    at Python/ceval.c:524
#89 0x00007f6041c6f6a2 in run_mod (arena=0x7f60420730a8, flags=0x7ffd6d021eb0,
    locals=0x7f6040d311f8, globals=0x7f6040d311f8, filename=0x7f6040c92858,
    mod=0x556a3c994740) at Python/pythonrun.c:1035
#90 PyRun_FileExFlags (fp=fp@entry=0x556a3c9747a0,
    filename_str=filename_str@entry=0x7f6040de96a8 "/job_exporter/main.py",
    start=start@entry=257, globals=globals@entry=0x7f6040d311f8,
    locals=locals@entry=0x7f6040d311f8, closeit=closeit@entry=1,
    flags=0x7ffd6d021eb0) at Python/pythonrun.c:988
#91 0x00007f6041c6f815 in PyRun_SimpleFileExFlags (fp=fp@entry=0x556a3c9747a0,
    filename=<optimized out>, closeit=closeit@entry=1,
    flags=flags@entry=0x7ffd6d021eb0) at Python/pythonrun.c:430
#92 0x00007f6041c6fc83 in PyRun_AnyFileExFlags (fp=fp@entry=0x556a3c9747a0,
    filename=<optimized out>, closeit=closeit@entry=1,
    flags=flags@entry=0x7ffd6d021eb0) at Python/pythonrun.c:84
#93 0x00007f6041c8f9f4 in pymain_run_file (p_cf=0x7ffd6d021eb0,
    filename=0x556a3c912b60 L"/job_exporter/main.py", fp=0x556a3c9747a0)
    at Modules/main.c:425
#94 pymain_run_filename (cf=0x7ffd6d021eb0, pymain=0x7ffd6d021fb0)
    at Modules/main.c:1520
#95 pymain_run_python (pymain=0x7ffd6d021fb0) at Modules/main.c:2520
#96 pymain_main (pymain=pymain@entry=0x7ffd6d021fb0) at Modules/main.c:2662
#97 0x00007f6041c8feaa in _Py_UnixMain (argc=<optimized out>,
    argv=<optimized out>) at Modules/main.c:2697
#98 0x00007f6040e102e1 in __libc_start_main (main=0x556a3aa587a0 <main>,
    argc=8, argv=0x7ffd6d0221d8, init=<optimized out>, fini=<optimized out>,
    rtld_fini=<optimized out>, stack_end=0x7ffd6d0221c8)
    at ../csu/libc-start.c:291
#99 0x0000556a3aa587da in _start ()
```

The \#25 entry is allocating for dict, which triggered GC and \#19 invoked gc callback, and then hanging in the \#1 entry waiting for a lock.

We found the issue when we are releasing for 0.9, and could not found root cause for this, so we used a temporary solution: restart whenever deadlock happens. We will solve this problem in long term.

### PR DESCRIPTION
**Pull request by xudifsd**
fix #2104 by disable GCCollector. Use twisted as http server. Also add signal handler for dumping stack trace to facilitate future debug for hanging.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -12,11 +12,11 @@ matrix:
         - python -m unittest discover test/
 
     - language: python
-      python: 2.7
```
Comment (fanyangCS) [inline .travis.yml]: I am not sure you can do that. please check with @ydye who may have some python 2.7 code.

File: src/job-exporter/deploy/job-exporter.yaml.template [hunk 2]
```diff
@@ -36,11 +36,11 @@ spec:
       containers:
       - image: {{ cluster_cfg["cluster"]["docker-registry"]["prefix"] }}job-exporter:{{ cluster_cfg["cluster"]["docker-registry"]["tag"] }}
         imagePullPolicy: Always
-        livenessProbe: # we found python may have deadlock and http server can not accept new conn, so use liveness probe to check if it is alive, and kill if it has deadlock. Temporary solution
```
Comment (fanyangCS) [inline src/job-exporter/deploy/job-exporter.yaml.template]: I think livenessProbe is a good idea. why don't we keep it as it is?

File: .travis.yml (hunk 1 above)
Comment (xudifsd) [reply .travis.yml]: This is test cases for watchdog, no unit tests from @ydye involved.

File: src/job-exporter/deploy/job-exporter.yaml.template (hunk 2 above)
Comment (xudifsd) [reply src/job-exporter/deploy/job-exporter.yaml.template]: Sure, then only livenessProbe, no readinessProbe.

**Comment by fanyangCS** [comment]
https://github.com/Microsoft/pai/pull/2116

**Pull request merged**